Atomic move

manast opened the pull request:
Implements #480 and #190

bradvogel commented on lib/job.js:
```diff
@@ -71,80 +70,65 @@ Job.fromId = function(queue, jobId){
-    if (lock){
-      _this.lock = lock;
-    } 
+Job.prototype.takeLock = function(){
```
This looks like it can be removed - right?

manast replied:
It is only used by a unit test indirectly, will try to get rid of it.

bradvogel replied:
Yeah, would be nice if the unit test implemented it itself. Just to keep the core queue implementation pure and simple.

manast replied:
yes, it is not tested, but used in some unit test where we need to have a lock: https://github.com/OptimalBits/bull/blob/v3.0.0/test/test_job.js#L100

bradvogel commented:
Looks like there's a lot of dead code that can be removed now:
`Job.prototype.move`
`Job.prototype.takeLock`
`Job.prototype.lockKey`
`Job.prototype.releaseLock`
`Job.prototype.delayIfNeeded`
`Job.prototype.moveToCompleted`
`scripts.move`

lots of others... feels good to clean it up!

bradvogel commented on lib/scripts/moveToActive.lua:
```diff
@@ -0,0 +1,42 @@
+end
+
+--[[
+  Release lock:
```
What was this for?

manast replied:
old stuff. removed now.

bradvogel commented on lib/queue.js:
```diff
@@ -726,65 +725,27 @@ Queue.prototype.processJob = function(job){
-    }, function(err){
-      _this.emit('error', err, 'failed to re-obtain lock before moving to failed, bailing');
+    //  job.takeLock(true /* renew */, false /* ensureActive */).then( function(/*lock*/) {
+    return job.moveToFailed(err).then(function(){
```
I think there's a hazard here: the lock can expire immediately after `stopTimer` is called and before the job is moved to failed. If `moveUnlockedJobsToWait` happens to run then, it'll reprocess the job. That's why we previously called `job.takeLock` right before calling job.moveToFailed.

To fix this, I think `stopTimer` should be moved to be called in a `.finally()` block after `moveToFailed`. That way `moveToFailed` can assume it has a lock.

manast replied:
I do not think this could happen in practice since the lock extender timer executes twice as often as the lock expiration time. However, if I move it to finally it also passes the unit tests, so I can move it there. Ultimately I will also make moveToFailed atomic, to further minimize any possible hazards.

bradvogel commented on lib/job.js:
```diff
@@ -57,7 +56,7 @@ Job.create = function(queue, name, data, opts){
 
   return addJob(queue, job).then(function(jobId){
-    job.jobId = jobId;
+    job.id = jobId;
```
This API change should be noted in the changelog

manast replied:
done.

bradvogel commented on lib/queue.js:
```diff
@@ -148,18 +151,13 @@ var Queue = function Queue(name, redisPort, redisHost, redisOptions){
   //
-  // Create blocking client (used to wait for jobs)
-  //
-  this.bclient = createClient('block');
```
One less connection - nice!

bradvogel commented on lib/queue.js:
```diff
@@ -268,14 +277,14 @@ var Queue = function Queue(name, redisPort, redisHost, redisOptions){
 
 util.inherits(Queue, Disturbed);
 
+Queue.ErrorMessages = errors.Messages;
```
Note this in the changelog?

bradvogel commented on lib/queue.js:
```diff
@@ -284,11 +293,11 @@ Queue.prototype.isReady = function(){
+  return this.client.commandQueue.length;
 };
 
 Queue.prototype.whenCurrentMoveFinished = function(){
```
What was this for originally? Seems like it can be removed to reduce complexity.

manast replied:
I have removed it.

bradvogel commented on lib/queue.js:
```diff
@@ -284,11 +293,11 @@ Queue.prototype.isReady = function(){
 }
 
 Queue.prototype.getJobMoveCount = function(){
```
What was this for originally? Seems like it can be removed to reduce complexity.

bradvogel commented on lib/scripts.js:
```diff
@@ -301,200 +213,66 @@ var scripts = {
    *    back to wait to be re-processed. To prevent jobs from cycling endlessly between active and wait,
    *    (e.g. if the job handler keeps crashing), we limit the number stalled job recoveries to MAX_STALLED_JOB_COUNT.
-
-   *  Case B) The job was just moved to 'active' from 'wait' and the worker that moved it hasn't gotten
```
You might want to remove `This is the most common case` above since it's now the only case.

manast replied:
yes, this has been removed.

bradvogel commented on lib/scripts.js:
```diff
@@ -584,131 +377,104 @@ var scripts = {
+  //
+  // TODO: DEPRECATE.
+  //
+  move: function(job, src, target){
```
I don't think this is used anymore - remove?

manast replied:
Only used in a unit test. will remove it.

manast commented on lib/scripts/moveToFinished.lua:
```diff
@@ -0,0 +1,56 @@
+
+     Output:
+      0 OK
+      1 Missing key.
```
fixed.

manast commented:
I pushed some updated addressing the code review. Regarding performance, I do not think it will affect in any measurable manner. If you have a busy queue, the publishing of 'added' event will not be used since the worker will just pick new jobs directly after processing the previous one. If not busy, the extra overhead of the publish will be negligible. Besides, in the previous blocking version we also published similar events, so the overhead was already there, and the blocking call was timed out after 2.5 seconds, effectively making it a polling queue. Overall the queue is much faster now, I will perform some benchmarks soon. The last thing remaining is a guard timmer and logic for handling the queue properly after a reconnection, we need better unit tests for that.

bradvogel commented on lib/queue.js:
```diff
@@ -678,113 +682,65 @@ Queue.prototype.processJob = function(job){
-    }, function(err){
-      _this.emit('error', err, 'failed to re-obtain lock before moving to failed, bailing');
-    });
+    //  job.takeLock(true /* renew */, false /* ensureActive */).then( function(/*lock*/) {
```
remove commented out line

bradvogel commented:
Yes- makes sense. Nice work! Excited for this.

manast commented:
with these latest commits I feel ready for merging in 3.0.0 branch and I would like to release the first alpha version for the 3.x series.

bradvogel commented:
Looks good to me to merge

chuym commented:
FWIW with respect to event listeners approach - it is actually not as CPU intensive as a polling alternative, if anything, it can be a source of potential memory leaks when having an unbound number of event listeners, nodejs avoids this by having a default max limit of event listeners of 10 to prevent a potential memory leak. The `EventEmitter` class is actually pretty straightforward, here: https://github.com/nodejs/node/blob/master/lib/events.js

manast commented on lib/commands/addJob-7.lua:
```diff
@@ -0,0 +1,99 @@
+    end
+  end
+
+  redis.call("PUBLISH", KEYS[4], jobId)
```
good catch. I will fix it.

manast commented on lib/commands/moveUnlockedJobsToWait-4.lua:
```diff
@@ -0,0 +1,54 @@
+       back to wait to be re-processed. To prevent jobs from cycling endlessly between active and wait,
+       (e.g. if the job handler keeps crashing), we limit the number stalled job recoveries to MAX_STALLED_JOB_COUNT.
+
+    DEPRECATED CASE:
```
I will remove the comment. With single instance redis, this case could not happen anymore since the lock is obtained atomically with the move. In the future, if we want to support redis with replication, we need to re-introduce redlock, but we will be able to take the lock *before* moving the job, so in practice the case should not happen either.

manast commented on lib/job.js:
```diff
@@ -22,20 +26,15 @@ var Job = function(queue, name, data, opts){
   }
 
-  opts = opts || {};
+  this.opts = _.extend({}, opts);
```
I will update using _.defaults since it is also semantically better. I feel options should always be kept in the opts field only, so delay should be in opts instead of a property of job, however, delay can also be defined when retrying a job, and in that case it would not be an option but a property used by the queue mechanics. Timestamp is also a bit special since normally it should be an internal property, but  sometimes you want to override it, just as job.id.

manast commented on lib/job.js:
```diff
@@ -71,80 +71,64 @@ Job.fromId = function(queue, jobId){
   };
 };
 
+Job.prototype.toData = function(){
```
yes, a nice simplification.

manast commented on lib/commands/moveToFinished-3.lua:
```diff
@@ -0,0 +1,56 @@
+
+     Output:
+      0 OK
+      -1 Missing key.
```
you are right. I added a -2 error code when a lock is missing. I noticed however that this error codes are not being used, I need to fix it as well, it may be hidding errors right now.

manast commented on lib/commands/moveToFinished-3.lua:
```diff
@@ -0,0 +1,56 @@
+
+if redis.call("EXISTS", KEYS[3]) == 1 then -- // Make sure job exists  
+  if ARGV[5] ~= "0" then
+    local lockKey = KEYS[3] .. ':lock'
```
I will add issue for such an improvement.

manast commented on CHANGELOG.md:
```diff
@@ -1,3 +1,10 @@
+
+- job.jobId changed to job.id.
+- refactored error messages into separate error module.
+- completed and failed job states are now represented in ZSETs.
```
- job.jobId -> job.id  to reduce redundancy.
- error messages refactored so that we can unit test without needing to copy/paste the messages.
- ZSETs allows us to efficiently get ranges of completed and failed jobs, very useful for building UIs and scripts.

manast commented:
@chuym thanks for the review I will commit the fixes directly to the 3.0.0 branch.
